File: 04_sorting/04_04_merge_sort/merge_sort.cpp
#include <bits/stdc++.h>

using namespace std;

void printArray(vector<int> &arr, int n)
{
    for (size_t i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
}

void merge(vector<int> &arr, int low, int mid, int high)
{
    vector<int> aux;

    int left = low;
    int right = mid + 1;

    while (left <= mid && right <= high)
    {
        if (arr[left] <= arr[right])
        {
            aux.push_back(arr[left]);
            left++;
        }
        else
        {
            aux.push_back(arr[right]);
            right++;
        }
    }

    while (left <= mid)
    {
        aux.push_back(arr[left]);
        left++;
    }

    while (right <= high)
    {
        aux.push_back(arr[right]);
        right++;
    }

    for (size_t i = low; i <= high; i++)
    {
        arr[i] = aux[i - low];
    }
    
}

void mergeSort(vector<int> &arr, int low, int high)
{
    if (low >= high) {
        return;
    }
    int mid = (low + high) / 2;
    mergeSort(arr, low, mid);
    mergeSort(arr, mid + 1, high);
    merge(arr, low, mid, high);
}

int main()
{
    vector<int> arr = {13, 46, 24, 52, 20, 9};
    int n = 6;

    cout << "Before:-" << endl;
    printArray(arr, n);
    mergeSort(arr, 0, n - 1);
    cout << endl;
    cout << "After:-" << endl;
    printArray(arr, n);
}

File: 04_sorting/04_04_merge_sort/merge_sort.py
def merge(arr: list, low: int, mid: int, high: int):
    left = low
    right = mid+1
    aux = []
    
    while left <= mid and right <= high:
        if arr[left] <= arr[right]:
            aux.append(arr[left])
            left += 1
        else:
            aux.append(arr[right])
            right += 1
    
    while left <= mid:
        aux.append(arr[left])
        left += 1
        
    while right <= high:
        aux.append(arr[right])
        right += 1
        
    for i in range(low, high + 1):
        arr[i] = aux[i-low]
        

def mergeSort(arr: list, low: int, high: int):
    if low < high:
        mid = (low + high) // 2
        mergeSort(arr, low, mid)
        mergeSort(arr, mid + 1, high)
        merge(arr, low, mid, high)
    

elements = [3, 2, 8, 5, 1, 4, 23]
print("Before:-")
print(elements)
mergeSort(elements, 0, len(elements) - 1)
print("After:-")
print(elements)